File: python/qa_indicators/qa_indicators/bollinger.py
from __future__ import annotations

import pandas as pd

from qa_core.schemas import IndicatorMeta, IndicatorOutput, ParamKind, ParamSpec

BOLLINGER_META = IndicatorMeta(
    id="bollinger",
    name_zh="布林带",
    name_en="Bollinger Bands",
    category="overlay",
    formula_tex=(
        r"\mathrm{Upper}_t = \mathrm{SMA}_n(P_t) + k\sigma_n,"
        r"\quad \mathrm{Lower}_t = \mathrm{SMA}_n(P_t) - k\sigma_n"
    ),
    params={
        "period": ParamSpec(kind=ParamKind.INT, default=20, min=2, max=400, step=1),
        "stddev": ParamSpec(kind=ParamKind.FLOAT, default=2.0, min=0.5, max=5.0, step=0.1),
        "source": ParamSpec(
            kind=ParamKind.ENUM,
            default="close",
            options=["open", "high", "low", "close"],
        ),
    },
    outputs=[
        IndicatorOutput(name="middle", kind="overlay"),
        IndicatorOutput(name="upper", kind="overlay"),
        IndicatorOutput(name="lower", kind="overlay"),
    ],
    use_cases_keys=["indicators.bollinger.use.volatility", "indicators.bollinger.use.meanrev"],
    pitfalls_keys=["indicators.bollinger.pit.trending"],
    references=[],
)


def compute(
    df: pd.DataFrame,
    period: int = 20,
    stddev: float = 2.0,
    source: str = "close",
) -> pd.DataFrame:
    """Compute Bollinger Bands (middle/upper/lower) of ``df[source]``.

    ``middle`` is the SMA; ``upper`` and ``lower`` are ``middle +/- stddev * sigma``
    where ``sigma`` is the population rolling standard deviation (``ddof=0``),
    matching ``pandas-ta`` defaults.
    """

    if period < 2:
        raise ValueError("period must be >= 2")
    if stddev <= 0:
        raise ValueError("stddev must be > 0")
    if source not in {"open", "high", "low", "close"}:
        raise ValueError(f"unsupported source: {source}")

    series = df[source].astype(float)
    middle = series.rolling(window=period, min_periods=period).mean()
    sigma = series.rolling(window=period, min_periods=period).std(ddof=0)
    upper = middle + stddev * sigma
    lower = middle - stddev * sigma

    return pd.DataFrame(
        {
            "middle": middle.to_numpy(),
            "upper": upper.to_numpy(),
            "lower": lower.to_numpy(),
        },
        index=df.index,
    )

File: packages/indicators-ts/src/bollinger.ts
import type { IndicatorInput, IndicatorOutput, Source } from './types'
import { coerceIntParam, coerceNumberParam, coerceSourceParam, pickSource } from './util'

/**
 * Bollinger Bands.
 *
 * - `middle` = SMA(period) of the source series.
 * - For every index where `middle` is defined, σ is the population (ddof=0)
 *   standard deviation of the same window. `upper = middle + stddev · σ`,
 *   `lower = middle - stddev · σ`.
 * - Population stddev is what pandas' `Series.rolling(window).std(ddof=0)`
 *   returns and matches numpy's default for ddof=0.
 * - The width-multiplier param is named `stddev` to match the Python
 *   canonical (`qa_indicators.bollinger.compute(..., stddev=2.0)`) and the
 *   golden parity JSON. The legacy name `k` is not honored.
 */
export function compute(input: IndicatorInput): IndicatorOutput {
  const period = coerceIntParam(input.params, 'period', 20)
  if (period < 2) throw new Error('period must be >= 2')
  const stddev = coerceNumberParam(input.params, 'stddev', 2)
  const source: Source = coerceSourceParam(input.params, 'source', 'close')
  const series = pickSource(input.bars, source)
  const n = series.length

  const middle = new Array<number | null>(n)
  const upper = new Array<number | null>(n)
  const lower = new Array<number | null>(n)
  for (let i = 0; i < n; i++) {
    middle[i] = null
    upper[i] = null
    lower[i] = null
  }
  if (n < period) return { middle, upper, lower }

  // Initial window sum + sum of squares.
  let sum = 0
  let sumSq = 0
  for (let i = 0; i < period; i++) {
    const v = series[i]
    if (v === undefined) throw new Error(`series[${i}] undefined`)
    sum += v
    sumSq += v * v
  }
  const seedMean = sum / period
  // Population variance: E[x^2] - (E[x])^2, clamp tiny negative drift to 0.
  const seedVar = Math.max(0, sumSq / period - seedMean * seedMean)
  const seedSigma = Math.sqrt(seedVar)
  middle[period - 1] = seedMean
  upper[period - 1] = seedMean + stddev * seedSigma
  lower[period - 1] = seedMean - stddev * seedSigma

  for (let i = period; i < n; i++) {
    const incoming = series[i]
    const outgoing = series[i - period]
    if (incoming === undefined || outgoing === undefined) {
      throw new Error(`series index ${i} out of range`)
    }
    sum += incoming - outgoing
    sumSq += incoming * incoming - outgoing * outgoing
    const mean = sum / period
    const variance = Math.max(0, sumSq / period - mean * mean)
    const sigma = Math.sqrt(variance)
    middle[i] = mean
    upper[i] = mean + stddev * sigma
    lower[i] = mean - stddev * sigma
  }

  return { middle, upper, lower }
}
